aptos-move/framework/aptos-framework/sources/permissioned_signer.move

```diff
@@ -0,0 +1,213 @@
+        capacity: u256,
+        perm: PermKey
+    ) acquires PermStorage {
+        assert!(
+            is_permissioned_signer(permissioned) &&
+            !is_permissioned_signer(master) &&
+            address_of(master) == address_of(permissioned),
+            error::permission_denied(ECANNOT_AUTHORIZE)
+        );
+        let permission_signer = permission_signer(permissioned);
+        let permission_signer_addr = address_of(&permission_signer);
+        if(!exists<PermStorage>(permission_signer_addr)) {
+            move_to(&permission_signer, PermStorage { perms: smart_table::new()});
+        };
+        let perms = &mut borrow_global_mut<PermStorage>(permission_signer_addr).perms;
+        smart_table::add(perms, copyable_any::pack(perm), capacity);
+    }
+    /// Asserts that the given signer has permission `PermKey`, and the capacity
+    /// to handle `weight`, which will be subtracted from capacity.
+    public fun check_permission<PermKey: copy + drop + store>(
+        s: &signer,
+        weight: u256,
+        perm: PermKey
+    ): bool acquires PermStorage {
+        if (!is_permissioned_signer(s)) {
+            // master signer has all permissions
+            return true
+        };
+        let addr = address_of(&permission_signer(s));
+        if(!exists<PermStorage>(addr)) {
+            return false
+        };
+        let perm = smart_table::borrow_mut(&mut borrow_global_mut<PermStorage>(addr).perms, copyable_any::pack(perm));
+        if(*perm < weight) {
+            return false
+        };
+        *perm = *perm - weight;
+        return true
+    }
+
+    public fun capacity<PermKey: copy + drop + store>(s: &signer, perm: PermKey): Option<u256> acquires PermStorage {
+        assert!(is_permissioned_signer(s), error::permission_denied(ENOT_PERMISSIONED_SIGNER));
+        let addr = address_of(&permission_signer(s));
+        if(!exists<PermStorage>(addr)) {
+            return option::none()
+        };
+        let perm_storage = &borrow_global<PermStorage>(addr).perms;
+        let key = copyable_any::pack(perm);
+        if(smart_table::contains(perm_storage, key)) {
+            option::some(*smart_table::borrow(&borrow_global<PermStorage>(addr).perms, key))
+        } else {
+            option::none()
+        }
+    }
+
+    public fun revoke_permission<PermKey: copy + drop + store>(permissioned: &signer, perm: PermKey) acquires PermStorage {
+        if(!is_permissioned_signer(permissioned)) {
+            // Master signer has no permissions associated with it.
+            return
+        };
+        let addr = address_of(&permission_signer(permissioned));
+        if(!exists<PermStorage>(addr)) {
+            return
+        };
+        smart_table::remove(&mut borrow_global_mut<PermStorage>(addr).perms, copyable_any::pack(perm));
+    }
+
+    /// Another flavor of api to extract and store permissions
+    public fun extract_permission<PermKey: copy + drop + store>(
```

Comment:
cc @igor-aptos

Reply:
maybe we should have a flavor where you can create permission signer from a vector of Permission<PermKey>, as well as have functions to grant Permission<PermKey> directly - instead of storing it?


for completeness - we should also have a way to return back what is left from the Permission<PermKey>